File: backend/src/jobhunter/application/candidate/preferences.py
"""Atomic configuration publication; no collection or external effects."""

from collections.abc import Callable
from uuid import uuid4

from sqlalchemy.engine import Connection

from jobhunter.application.manual_application_entries.service import utc_now
from jobhunter.domain.preferences.models import (
    Configured,
    NotConfigured,
    PreferenceSet,
    PreferenceSetVersion,
    SavePreferences,
    SaveResult,
    fingerprint,
)
from jobhunter.domain.shared.errors import Failure
from jobhunter.domain.shared.values import MAX_REVISION
from jobhunter.infrastructure.persistence.sqlalchemy.repositories.preferences import (
    PreferenceRepository,
)
from jobhunter.infrastructure.persistence.sqlalchemy.uow.store import Store


class Preferences:
    def __init__(self, store: Store, *, clock: Callable[[], str] = utc_now) -> None:
        self.store = store
        self.clock = clock

    def current(self) -> Configured | NotConfigured:
        def read(conn: Connection) -> Configured | NotConfigured:
            current = PreferenceRepository(conn).current()
            if current is None:
                return NotConfigured(status="NOT_CONFIGURED")
            return Configured(
                status="CONFIGURED",
                preference_set=current[0],
                current_preference_set_version=current[1],
            )

        return self.store.run(read)

    def version(self, identity: str) -> PreferenceSetVersion:
        return self.store.run(lambda conn: PreferenceRepository(conn).version(identity))

    def save(self, command: SavePreferences) -> SaveResult:
        # Re-admit even application callers; model_construct/mutated lists cannot bypass rules.
        command = SavePreferences.model_validate(command.model_dump())
        digest = fingerprint(command)

        def execute(conn: Connection) -> SaveResult:
            repo = PreferenceRepository(conn)
            receipt = repo.receipt(command.request_id, digest)
            if receipt is not None:
                return receipt
            current = repo.current()
            if (current is None and command.revision is not None) or (
                current is not None and command.revision != current[0].revision
            ):
                raise Failure("REVISION_CONFLICT")
            if current is not None and current[1].configuration == command.configuration:
                result = SaveResult(
                    preference_set_id=current[0].preference_set_id,
                    preference_set_version_id=current[1].preference_set_version_id,
                    revision=current[0].revision,
                    outcome="UNCHANGED",
                )
            else:
                if current is not None and current[0].revision == MAX_REVISION:
                    raise Failure("REVISION_EXHAUSTED")
                now = self.clock() if current is None else max(self.clock(), current[0].updated_at)
                root = PreferenceSet(
                    preference_set_id=str(uuid4())
                    if current is None
                    else current[0].preference_set_id,
                    current_preference_set_version_id=str(uuid4()),
                    revision=1 if current is None else current[0].revision + 1,
                    created_at=now if current is None else current[0].created_at,
                    updated_at=now,
                )
                version = PreferenceSetVersion(
                    preference_set_version_id=root.current_preference_set_version_id,
                    preference_set_id=root.preference_set_id,
                    created_at=now,
                    configuration=command.configuration,
                )
                repo.publish(root, version, first=current is None)
                self.store.fault("preferences_after_publication")
                result = SaveResult(
                    preference_set_id=root.preference_set_id,
                    preference_set_version_id=version.preference_set_version_id,
                    revision=root.revision,
                    outcome="CREATED" if current is None else "UPDATED",
                )
            repo.record(command.request_id, digest, result)
            self.store.fault("preferences_after_receipt")
            return result

        return self.store.run(execute, write=True)

File: frontend/tests/fixtures/preferences.ts
import type {
  Configuration,
  CurrentPreferences,
  SaveResult,
} from '@/features/preferences/model';
const root = '22222222-2222-4222-8222-222222222222';
const version = '33333333-3333-4333-8333-333333333333';
export const configuration: Configuration = {
  target_job_keywords: ['工程师'],
  accepted_cities: { mode: 'UNLIMITED' },
  minimum_salary: { mode: 'LIMITED', value: 15000 },
  recruitment_types: { mode: 'UNLIMITED' },
  excluded_companies: { mode: 'UNLIMITED' },
  max_required_education: { mode: 'UNLIMITED' },
};
export function current(
  revision = 1,
  keywords = ['工程师'],
): CurrentPreferences {
  return {
    status: 'CONFIGURED',
    preference_set: {
      preference_set_id: root,
      current_preference_set_version_id: version,
      revision,
      created_at: '2026-09-20T01:00:00.000Z',
      updated_at: '2026-09-20T01:00:00.000Z',
    },
    current_preference_set_version: {
      preference_set_id: root,
      preference_set_version_id: version,
      created_at: '2026-09-20T01:00:00.000Z',
      configuration: { ...configuration, target_job_keywords: keywords },
    },
  };
}
export const saved: SaveResult = {
  preference_set_id: root,
  preference_set_version_id: version,
  revision: 2,
  outcome: 'UPDATED',
};
